File: sdk/python/tap/x402/response.py
"""x402 `X-PAYMENT-RESPONSE` for the TAP `tap.v1.channel` scheme.

Returned by the producer once the channel-open transaction confirms. The
consumer treats this as the channel-open acknowledgement and proceeds to
the streaming phase."""

from __future__ import annotations

import base64
import json
from dataclasses import dataclass
from typing import Any

from tap.exceptions import X402Error


@dataclass(frozen=True, slots=True)
class PaymentResponse:
    tx_hash: str
    settlement: str  # "confirmed" | "finalized"
    channel_id: str
    channel_state: str  # "active"

    def to_payload(self) -> dict[str, Any]:
        return {
            "tx_hash": self.tx_hash,
            "settlement": self.settlement,
            "extra": {
                "channel_id": self.channel_id,
                "channel_state": self.channel_state,
            },
        }


def encode_response(resp: PaymentResponse) -> str:
    raw = json.dumps(resp.to_payload(), separators=(",", ":")).encode("utf-8")
    return base64.b64encode(raw).decode("ascii")


def decode_response(header_value: str) -> PaymentResponse:
    try:
        payload = json.loads(base64.b64decode(header_value, validate=True))
    except (ValueError, json.JSONDecodeError) as exc:
        raise X402Error("X-PAYMENT-RESPONSE is not valid base64-JSON") from exc

    extra = payload.get("extra") or {}
    try:
        return PaymentResponse(
            tx_hash=str(payload["tx_hash"]),
            settlement=str(payload["settlement"]),
            channel_id=str(extra["channel_id"]),
            channel_state=str(extra["channel_state"]),
        )
    except (KeyError, ValueError, TypeError) as exc:
        raise X402Error("X-PAYMENT-RESPONSE is missing or malformed fields") from exc

File: sdk/typescript/src/x402/response.ts
/**
 * x402 `X-PAYMENT-RESPONSE` for the TAP `tap.v1.channel` scheme. Returned by
 * the producer once the channel-open transaction confirms; the consumer
 * treats this as the channel-open acknowledgement and proceeds to the
 * streaming phase. Mirrors `tap.x402.response`.
 */

import { X402Error } from "../exceptions.js";
import { _internal } from "../protocol/codec.js";

export interface PaymentResponse {
  readonly txHash: string;
  /** "confirmed" | "finalized" */
  readonly settlement: string;
  readonly channelId: string;
  /** "active" — channel lifecycle state at ack time. */
  readonly channelState: string;
}

interface RootPayload {
  tx_hash: string;
  settlement: string;
  extra: {
    channel_id: string;
    channel_state: string;
  };
}

function toPayload(r: PaymentResponse): RootPayload {
  return {
    tx_hash: r.txHash,
    settlement: r.settlement,
    extra: {
      channel_id: r.channelId,
      channel_state: r.channelState,
    },
  };
}

export function encodeResponse(resp: PaymentResponse): string {
  return _internal.utf8ToBase64(JSON.stringify(toPayload(resp)));
}

export function decodeResponse(headerValue: string): PaymentResponse {
  let payload: RootPayload;
  try {
    payload = JSON.parse(_internal.base64ToUtf8(headerValue)) as RootPayload;
  } catch (cause) {
    throw new X402Error("X-PAYMENT-RESPONSE is not valid base64-JSON", {
      cause: cause as Error,
    });
  }
  const extra = payload.extra;
  if (!extra) {
    throw new X402Error("X-PAYMENT-RESPONSE is missing or malformed fields");
  }
  try {
    return {
      txHash: String(payload.tx_hash),
      settlement: String(payload.settlement),
      channelId: String(extra.channel_id),
      channelState: String(extra.channel_state),
    };
  } catch (cause) {
    throw new X402Error("X-PAYMENT-RESPONSE is missing or malformed fields", {
      cause: cause as Error,
    });
  }
}
